### PR TITLE
feat : new dp algo added `UniqueSubsequenceCount.java`

### PR DESCRIPTION
**Pull request by Tuhinm2002**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Comment by Tuhinm2002** [comment]
@siriak @alxkm  Added new algo in dp. Please review it

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    public static int subseqCount(String str) {
+
+        int[] dp = new int[str.length() + 1];
+        Arrays.fill(dp, -1);
+
+        return recursiveCall(str, 0, dp);
+    }
+
+    public static int recursiveCall(String st, int idx, int[] dp) {
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:25]: What do you think to rename method to `countSubsequences`?

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.dynamicprogramming;
+
+// Program to find the number of Subsequences can be produced from a string
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:3]: This comment should be formed as in standard javadoc and removed from here

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.dynamicprogramming;
+
+// Program to find the number of Subsequences can be produced from a string
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+// Author -> https://github.com/Tuhinm2002
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:9]: This should be in javadoc

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.dynamicprogramming;
+
+// Program to find the number of Subsequences can be produced from a string
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+// Author -> https://github.com/Tuhinm2002
+
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:10]: This new line should be removed

File: src/test/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCountTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.thealgorithms.dynamicprogramming;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public final class UniqueSubsequencesCountTest {
+
+    @Test
+    void subseqCountTestOne() {
```
Comment (alxkm) [inline src/test/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCountTest.java:10]: Better to refactor using parametrized tests (a lot of examples already exist in project).

And will be good to have more test cases. Possible testcases:

```
Empty String:
Should return 0 when the input string is empty.

Single Character:
A string with only one character should have one subsequence (the character itself).
"a"

Two Different Characters:
For a string with two different characters, there should be 3 subsequences: a, b, and ab.
"ab"

String with Repeated Characters:
A string with repeated characters should handle duplicates correctly.
"aaa"

All Unique Characters:
A string with all unique characters should have the maximum number of subsequences.
"abc"

String with Mixed Characters:
A string with a mix of repeated and unique characters.
"abab"

A string containing non-alphabetical characters like numbers or punctuation.
"a1b2"

Long String with Repeated Characters:
Tests ability to handle long input strings with many repeated characters. (But looks like such test already exist)
"aaaaaa"

String with Spaces:
Ensure that spaces are treated like any other character in the subsequence count.
"a b"

String with Upper and Lower Case:
A string containing both uppercase and lowercase characters to test case sensitivity.
"AaBb"

```

**Review by alxkm — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java (hunk 1 above)
Comment (Tuhinm2002) [reply src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:62]: Actually it is named as per its duty so I kept it like that. But yeah your function perfect for initial function call which is written above

**Comment by Tuhinm2002** [comment]
@alxkm hey check the changes and review it

File: src/test/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCountTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Test
+    void subseqCountTestOne() {
+        String s = "abc";
+
+        assertEquals(7, UniqueSubsequencesCount.countSubsequences(s));
+    }
+
+    @Test
+    void subseqCountTestTwo() {
```
Comment (alxkm) [inline src/test/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCountTest.java:17]: The code not written via parameterized tests approach. Please you them. There are a lot of such examples in project.

File: src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java (hunk 1 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/dynamicprogramming/UniqueSubsequencesCount.java:62]: I understand that the current name reflects its role, but I believe using more descriptive names like `countSubsequences `helps clarify the method’s purpose at a glance. Many established algorithms like `Quick Sort` or `Merge Sort` don’t emphasize that they’re recursive, but rather focus on what the function achieves. In this case, `countSubsequences `explicitly tells us what the method does, which I think improves readability and maintainability.

**Review by alxkm — CHANGES_REQUESTED** [review]

**Comment by Tuhinm2002** [comment]
@alxkm check the new code for review

**Review by alxkm — APPROVED** [review]

**Review by alxkm — APPROVED** [review]

**Pull request merged**